Initial Daikin S21 integration

u20p17 opened the pull request:
Daikin (S21) – Einstellungen, Funktionen & Hinweise

Folgende Einstellungen und Funktionen stehen für Daikin Klimageräte (S21-Schnittstelle am Innengerät) zur Verfügung.

Hinweis: Das S21-Protokoll ist nicht vollständig dokumentiert und einige Funktionen sind je nach Gerätemodell / Protokoll-Version optional. Die Implementierung nutzt nur dokumentierte bzw. verifizierte Teile (Stand: [Faikin Wiki Protokoll v0–v3.40](https://github.com/revk/ESP32-Faikin/wiki/S21-Protocol)). Nicht unterstützte oder unsichere Bereiche wurden bewusst weggelassen.

Telemetrie & Überwachung
Robuste Kommunikation: Retry-Mechanismen und Timeout-Behandlung
Innengerät-Temperatur (RH)
Außentemperatur (Ra)
Feuchtigkeit (Re)
Gesamtverbrauch (FM)
Kompressorfrequenz (Rd)
Fan Speed (RL / RK): Drehzahl / Tap (wenn verfügbar)
Echter Zielwert (RX): Interner Regler-Zielwert (inkl. Powerful/Intelligent Eye Einflüsse)
Unit/System State (RzB2 / RzC3): Bitfelder für Aktiv / Defrost / Powerful etc.

Grundfunktionen
Ein/Aus: Steuerung über D1 (Power-Bit)
Betriebsmodi: Automatik, Kühlen, Heizen, Trocknen, Lüfter (gemäß F1-Rückmeldung)
Solltemperatur: 18,0–30,0 °C in 0,5 °C-Schritten
Lüftergeschwindigkeit: Automatik + 5 Stufen (Low → High). Quiet/Night intern als eigener Modus; Rückmeldung über RG korrekt, F1 zeigt Quiet wie Auto (Protokoll-Einschränkung).
Swing: Vertikal / Horizontal / Beides über D5, sofern vom Gerät gemeldet (F2-Featurebits). Keine absolute Lamellenpositions-Steuerung.

Erweiterte / optionale Funktionen
Die Verfügbarkeit wird zur Laufzeit über F2 / FK / (bei v3+) FU00 erkannt. Nicht jedes Gerät unterstützt alle Funktionen.
Powerful (F6 / D6): Wenn F6 unterstützt (ab vielen v2/v3-Geräten). Frühe Geräte evtl. nur Status über F3-Bitmuster.
Econo (F7 / D7): Energiesparmodus (Bit im zweiten Byte von F7).
Quiet (D4 proprietär in dieser Implementierung / Status in F6 Quiet-Bit). Fan Quiet wird durch RG korrekt gelesen; F1 meldet ihn als Auto (Protokoll-Eigenheit).
Streamer: Wenn in FU00 Byte 5 vorhanden und F6 Streamer-Bit reagiert.
LED: Steuerung (abhängig von LED-Bit in F6 Byte 3, invertierte Logik). Fehlt bei älteren Modellen.
Kein externer Temperatursensor via S21: F6 Sensor-Bit wählt keine externe Raumtemperaturquelle; es werden keine externen Temperaturdaten übergeben.
Demand / Leistungsvorgabe (F7): Auswertung des Rohwertes (1-stellige Lastkennzahl), Umsetzung derzeit heuristisch in Prozent (experimentell).

Hinweis: Eine eventuelle zukünftige Erweiterung würde deutlich gekennzeichnet und setzt belastbare Protokollnachweise voraus.

Protokollversionen
Dynamische Erkennung v0 / v2 / v3.x (F8 + FY00 Probe). Antwort „v2“ über F8 bei v3 ist Daikin-Rückwärtskompatibilität.

Automatische Funktionen
Protokoll-Detection: F8 → ggf. FY00 für v3+; Feature-Bits über F2/FK/F6/FU00
Polaritätshandling: Automatische Polarity-Versuche (Inversion RX/TX-Kombinationen)
Fallback F6→F3: Falls F6 (Special Modes) nicht reagiert, werden ggf. F3-Bits genutzt (eingeschränkt)
Retry / Timeout: ACK- und Grace-Window Handling (angepasst an beobachtete Zeiten)

Der Vorgabewert wird auf den nächstliegenden fest definierten Wert gerundet.

Getestete Geräte
Daikin Stylish FTXA25AW

Rückmeldungen aus Tests willkommen – unbekannte Feature-Kombinationen werden protokolliert und können zukünftige Erweiterungen steuern.

Faikin-Bezug
Konzepte (Timing, Query-Reihenfolge, Interpretation von F*/R*/Rz*-Kommandos) orientieren sich an den im Faikin-Projekt dokumentierten Erkenntnissen. Nicht jede dort aufgeführte experimentelle Funktion ist hier aktiv umgesetzt.

u20p17 committed:
WIP: Daikin S21 integration and local tweaks

mgeramb commented on src/AirConditionDriver.h:
```diff
@@ -52,6 +59,26 @@ class AirConditionDriverStatusFeedback
+    // --- Extended feature update interface (non-breaking: provide virtuals with default empty bodies) ---
+    // These allow drivers to push richer telemetry without forcing every implementation to override.
+    // If you have an implementation, override the ones you need.
+    virtual void updatePower(bool power) { powerChanged(power); }
```
Ich habe die Funktionen bewusst abstract gemacht, damit man im jeweiligen Treiber sieht, was eventuell noch implementiert werden sollte.

mgeramb commented on src/AirConditionDriver.h:
```diff
@@ -52,6 +59,26 @@ class AirConditionDriverStatusFeedback
+    virtual void updateOnlineStatus(bool /*online*/) { }
+    virtual void updateWifiLed(bool /*on*/) { }
+    virtual void updateHumidity(uint8_t /*humidity*/) { }
+    virtual void updateHumidityMode(uint8_t /*humidityMode*/) { }  // 0=Off, 1=Low, 2=Standard, 3=High, 4=Continuous
```
Fixe Werte vorgeben wie hier im Kommentar ist nicht gut, das kann von Modell zu Modell unterschiedlich sein. Da benötigt es eine Funktion die sagt, wieviele Stufe es gibt. In der ETS sollte es dann auf Prozent gemappt werden, damit es mit allen Modellen Funktioniert

mgeramb commented on src/Aircondition.share.xml:
```diff
@@ -815,6 +822,10 @@
                       <ComObjectRefRef RefId="%AID%_O-%TT%00060_R-%TT%0006001" />
                       <!-- Gerätemodus Silent 2 Status -->
                       <ComObjectRefRef RefId="%AID%_O-%TT%00061_R-%TT%0006101" />
+                      <!-- Luftfeuchtigkeits Status -->
```
Wir brauchen hier noch choose when Konstrukte um die ComObjectRefRef um die neuen für Toshiba unsichtbar zu machen

mgeramb commented:
Sehr coole Arbeit!
Inwieweit ist der Code schon mit allen Funktionen getestet? Mir wäre wichtig, das wir alles aus der ETS App mit Conditions rausnehmen, was derzeit mit Daikin noch nicht geht. Z.B. gehe ich davon aus, dass die Szenen nicht funktionieren. Umgekehrt sollten Funktionen die es bei Toshiba nicht gibt, dort ausgenommen werden, wie z.B. der neue KO für Luftfeuchtigkeit.

mgeramb reviewed (changes requested):
Hier ein paar Anmerkungen, an welchen Stellen noch Ändernungen notwendig sind

u20p17 committed:
make extended telemetry channels abstract. add getMaximumHumidityModeLevels() and calculate a percent value for Humidity mode

Missing: add KoAIR_HumidityModePercent and KoAIR_TotalEnergy_kWh + hide some fields in ETS depending on the Manufacturer

u20p17 committed:
add KOs for HumidityMode and TotalEnergy

u20p17 committed:
Show humidity, humidityMode and total Energy KOs only for Daikin; fixes typos

u20p17 committed:
Hide KOs for Daikin (Lamelle Vertikal Höher/Tiefer, Lamelle Horizontal Links/Rechts, Lamelle Vertikale Position , Lamelle Vertikale Position Status)

u20p17 committed:
hide 'Vertikale Lammellenstellung' and 'Leistungsbegrenzung' for Daikin in the scenes
